Update: rewrite TokenStore (fixes #7810)

mysticatea opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[X] Add something to the core

Fixes #7810

**What changes did you make? (Give an overview)**

I rewrote `TokenStore` with few extensions.

- We had been having 2 `TokenStore` instances. One is to access tokens. Another is to access a mix of tokens and comments. I added `{includeComments: true}` option to every methods of `TokenStore`. As the result, we no longer need the store of a mix of tokens and comments. This would reduce memory usage of linting.
- I added some enhancements to `TokenStore`. I think those enhancements are useful to skip parentheses or something like.
    - `getFirstTokenBetween(left, right, options)`
    - `getLastTokenBetween(left, right, options)`
    - `filter` option to skip specific tokens.
    - [token-store.d.ts](https://gist.github.com/mysticatea/b15211db22f2a318b04462f90c8f60da) is the interface of `TokenStore`.
 - I refactored our codebase with the enhancements. For examples:

```diff
// lib/rules/arrow-body-style.js
-                            const lastTokenBeforeBody = sourceCode.getTokensBetween(sourceCode.getFirstToken(node), arrowBody)
-                                .reverse()
-                                .find(token => token.value !== "(");
+                            const lastTokenBeforeBody = sourceCode.getLastTokenBetween(sourceCode.getFirstToken(node), arrowBody, astUtils.isNotOpeningParenToken);
```

```diff
// lib/rules/arrow-spacing.js
-            let arrow = sourceCode.getTokenBefore(node.body);
-
-            // skip '(' tokens.
-            while (arrow.value !== "=>") {
-                arrow = sourceCode.getTokenBefore(arrow);
-            }
+            const arrow = sourceCode.getTokenBefore(node.body, astUtils.isArrowToken);
```

```diff
// lib/rules/func-call-spacing.js
+            const lastToken = sourceCode.getLastToken(node);
             const lastCalleeToken = sourceCode.getLastToken(node.callee);
-            let prevToken = lastCalleeToken;
-            let parenToken = sourceCode.getTokenAfter(lastCalleeToken);
-
-            // advances to an open parenthesis.
-            while (
-                parenToken &&
-                parenToken.range[1] < node.range[1] &&
-                parenToken.value !== "("
-            ) {
-                prevToken = parenToken;
-                parenToken = sourceCode.getTokenAfter(parenToken);
-            }
+            const parenToken = sourceCode.getFirstTokenBetween(lastCalleeToken, lastToken, astUtils.isOpeningParenToken);
+            const prevToken = parenToken && sourceCode.getTokenBefore(parenToken);
```

**Is there anything you'd like reviewers to focus on?**

- Please check if the purpose of #7810 is satisfied or not.
- Please check the new interface.
- Please check the English of documents.

mysticatea commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+     */
+    constructor(tokens, comments) {
+        this.tokens = tokens;
+        this.comments = comments.slice(0); // Needs cloning for backward compatibility!
```
I noticed shebang's comment (`///usr/bin/env node`) remains in token store currently. I think this is a bug, but I keep it to avoid breaking changes.

> ※ `comments` needs cloning for backward compatibility.
> After this initialization, ESLint removes a shebang's comment from `comments`.
> However, so far we had been concatenating 'tokens' and 'comments',
> so the shebang's comment had remained in the concatenated array.
> As the result, both `getTokenOrCommentAfter` and `getTokenOrCommentBefore`
> methods had been returning the shebang's comment.
> And some rules depend on this behavior.

kaicataldo replied:
Sorry, I don't have time to do a full review at the moment, but [this PR](https://github.com/eslint/eslint/pull/7516) changes how we handle shebangs. It essentially creates a comment token type ("Shebang") and then treats it just like any other comment. So I think we can leave the behavior as is until that PR lands?

mysticatea replied:
> So I think we can leave the behavior as is until that PR lands?

Yes, I think so.
`Shebang` type comments sound nice to me.

ilyavolodin commented:
@not-an-aardvark Could you post perf number before and after this change please?

not-an-aardvark commented:
@ilyavolodin I'm confused, did you mean to ping @mysticatea instead?

ilyavolodin commented:
Oh, sorry, my bad. I mean to ping @mysticatea

not-an-aardvark commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+
+        // Choose the current token from the token or the comment.
+        const t = (this.it >= 0) ? this.tokens[this.it] : null;
+        const c = (this.ic >= 0) ? this.comments[this.ic] : null;
```
The variable names `t` and `c` are a bit confusing. Maybe they should be something like `previousToken` and `previousComment` instead.

not-an-aardvark commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+     * Initializes this cursor.
+     */
+    constructor() {
+        this.current = null;
```
I think it might be more readable if `this.current` was called `this.currentToken` (or maybe `this.currentValue`) instead. At first, I couldn't tell whether `this.current` was referring to an index.

mysticatea replied:
Actually, [IEnumerator](https://msdn.microsoft.com/en-us/library/78dfe2yb.aspx) was in my head :smiley:

not-an-aardvark commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+    /** @inheritdoc */
+    moveNext() {
+        while (super.moveNext()) {
+            if (this.predicate(this.current)) {
```
It's only a minor issue, but `predicate` will get called with a `this` value of the current instance of `FilterCursor`, which would be observable to external code. This could make it difficult to change the implementation in the future without making breaking changes to rules.

Maybe we should call the predicate as a pure function:

```js
const predicate = this.predicate;

while (super.moveNext()) {
    if (predicate(this.current)) {
        // ...
```

mysticatea replied:
Nice catch!

not-an-aardvark commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+    movePrev() {
+        while (this.count > 0) {
+            this.count -= 1;
+            if (!super.movePrev()) {
```
It looks like the `skip` behavior depends on whether `moveNext` or `movePrev` is called first. If `moveNext` is called first, then the *first* `count` tokens are skipped, but if `movePrev` is called first, then the *last* `count` tokens are skipped.

I'm not sure if there's an easy workaround for this, though.

mysticatea replied:
This is intentional.

not-an-aardvark commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+    moveNext() {
+        if (this.count > 0) {
+            this.count -= 1;
+            return super.moveNext();
```
It looks like after calling `moveNext` enough times, `movePrev` won't work at all. 

```js
for (let i = 0; i < count; i++) {
    this.moveNext();
}
this.movePrev(); // => false
```

I would expect that after using `movePrev`, the cursor would be at index `count - 2`.

mysticatea replied:
It's never used in a mix of `moveNext` and `movePrev`.
`getTokenAfter` uses `moveNext` only, `getTokenBefore` uses `movePrev` only.
So this is no problem.

mysticatea replied:
I removed `movePrev()` as it causes confusing.

not-an-aardvark commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+     * @protected
+     */
+    get firstIndex() {
+        if (typeof this.imap[this.start] !== "undefined") {
```
Since `this.imap` is created from `Object.create(null)`, this could also be

```js
if (this.start in this.imap) {
```

However, I'm fine with either method.

mysticatea replied:
I will fix it.
I thought `in` operator is slow since I had seen nodejs/node#10443. But this seems a different situation.
https://jsperf.com/undefined-check-777

not-an-aardvark replied:
Another option would be to use a `Map` to avoid prototype lookups.

mysticatea replied:
Actually, I had used `Map` as first, but it was 16% slower than this for getting value. :cry: 

https://jsperf.com/int-map

btmills commented on lib/ast-utils.js:
```diff
@@ -247,20 +247,213 @@ function isParenthesised(sourceCode, node) {
+ * @param {Token} token - The token to be check.
+ * @returns {boolean} `false` if the token is a comma token.
+ */
+function isNotCommaToken(token) {
```
Another possibility would be maintaining the current API but using the helper to pre-negate the positive definitions.

```js
function isCommaToken(token) {
    return token.value === "," && token.type === "Punctuator";
}

const isNotCommaToken = negate(isCommaToken);
```

With either solution, we get to avoid maintaining a set of implementations as well as their inverses separately.

mysticatea replied:
I will try it later.
I'm worried performance reducing by the dynamic function call since those helpers are called in a lot of time.

btmills replied:
If there ends up being a noticeable performance difference, I would agree leaving this as-is 👍

mysticatea commented:
Thank you for the review!
I will check those.

----

@ilyavolodin Sure.

I was careful about performance.
But sadly, the new implementation is about 1% slower than the original. Also after the refactoring of 	8f85577, it got 3% slower, I guess it was caused by that `FilterCursor` increases function calls of each iteration.

<details><summary>Open raw log</summary>

```
C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master ≡]> npm run perf

> eslint@3.13.1 perf C:\Users\t-nagashima.AD\Documents\GitHub\eslint
> node Makefile.js perf


Loading:
  Load performance Run #1:  237.425139ms
  Load performance Run #2:  239.108839ms
  Load performance Run #3:  232.480586ms
  Load performance Run #4:  237.605341ms
  Load performance Run #5:  235.970236ms

  Load Performance median:  237.425139ms


Single File:
  CPU Speed is 3392 with multiplier 13000000
  Performance Run #1:  3946.4067999999997ms
  Performance Run #2:  3938.73959ms
  Performance Run #3:  3994.417041ms
  Performance Run #4:  3911.2673059999997ms
  Performance Run #5:  3938.744118ms

  Performance budget exceeded: 3938.744118ms (limit: 3832.5471698113206ms)


Multi Files (0 files):
  CPU Speed is 3392 with multiplier 39000000
  Performance Run #1:  12499.78252ms
  Performance Run #2:  12556.26077ms
  Performance Run #3:  12439.634717ms
  Performance Run #4:  12459.592524ms
  Performance Run #5:  12317.274179ms

  Performance budget exceeded: 12459.592524ms (limit: 11497.641509433963ms)

C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master ≡]> git checkout rewrite-token-store
Switched to branch 'rewrite-token-store'
Your branch is up-to-date with 'origin/rewrite-token-store'.
C:\Users\t-nagashima.AD\Documents\GitHub\eslint [rewrite-token-store ≡]> npm run perf

> eslint@3.13.1 perf C:\Users\t-nagashima.AD\Documents\GitHub\eslint
> node Makefile.js perf


Loading:
  Load performance Run #1:  236.285969ms
  Load performance Run #2:  235.878777ms
  Load performance Run #3:  239.786486ms
  Load performance Run #4:  237.061413ms
  Load performance Run #5:  241.826668ms

  Load Performance median:  237.061413ms


Single File:
  CPU Speed is 3392 with multiplier 13000000
  Performance Run #1:  4109.42097ms
  Performance Run #2:  4072.425119ms
  Performance Run #3:  4087.458598ms
  Performance Run #4:  4223.57671ms
  Performance Run #5:  4177.791984ms

  Performance budget exceeded: 4109.42097ms (limit: 3832.5471698113206ms)


Multi Files (0 files):
  CPU Speed is 3392 with multiplier 39000000
  Performance Run #1:  13025.187907ms
  Performance Run #2:  13013.918458ms
  Performance Run #3:  12832.759673ms
  Performance Run #4:  12840.481519ms
  Performance Run #5:  12834.366706ms

  Performance budget exceeded: 12840.481519ms (limit: 11497.641509433963ms)
```

</details>

mysticatea commented:
@not-an-aardvark I love Iterators/Generators! Actually, I had written `TokenStore` with generators: https://gist.github.com/mysticatea/ff83097cdebb0aad79fe6cccc29d4f50. But unfortunately, it was super slow. `npm run perf` was 15% slower than this implementation.

mysticatea commented on lib/ast-utils.js:
```diff
@@ -247,20 +247,213 @@ function isParenthesised(sourceCode, node) {
+ * @returns {boolean} `true` if the token is a comment token.
+ */
+function isCommentToken(token) {
+    return token.type === "Line" || token.type === "Block";
```
Oh, nice catch!

mysticatea commented on lib/rules/keyword-spacing.js:
```diff
@@ -59,6 +59,18 @@ function isCloseParenOfTemplate(token) {
+ * @returns {boolean} `true` if the token is an `of` token.
+ */
+function isOfToken(token) {
+    return token.value === "of";
```
So, `for (const x of /*of*/ iterable) {}` triggers false positive. Wow, it's a bug in the current master as well. I will fix it.

mysticatea commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
-        length = tokens.length;
+const W_INIT = -3;
+const W_TOKENS = 1;
+const W_COMMENTS = 2;
```
Those are used `TokenCommentCursor#which` property; the flag which indicates the cursor is using which index. I will add comments.

mysticatea commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+ * @returns {number} The found index or `tokens.length`.
+ * @private
+ */
+function search(tokens, location) {
```
I will try it.

mysticatea commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+     * This consumes this cursor.
+     * @type {Token|Comment}
+     */
+    get firstToken() {
```
I will fix it.

mysticatea commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+        this.comments = comments;
+        this.imap = imap;
+        this.start = start | 0;
+        this.end = end | 0;
```
I will fix it.

This is vestige I fought for inlining of methods.

mysticatea commented on lib/token-store.js:
```diff
@@ -8,196 +8,1110 @@
+        cursor = new FilterCursor(cursor, filter);
+    }
+    if (skip >= 1) {
+        cursor = new SkipCursor(cursor, skip);
```
Yes, this is intentional.

platinumazure replied:
For what it's worth, I would find that unintuitive and would prefer that skipping "happens first" (i.e., skipping happens regardless of filtering, and then the filter takes over). Makes sense when you want to get a token using a predicate but first skip over a known bracket/paren/brace, etc.

mysticatea replied:
Hmm, I thought this is natural, for example, "I want get the 2nd comma" → `sourceCode.getFirstToken(node, {skip: 1, filter: isComma})`

platinumazure replied:
I guess it would be nice to be able to go either way on this. But I'm not sure I can provide any use case to back up my preference.

mysticatea replied:
Or, I think we can throw errors if both `skip` and `filter` are given. I didn't find cases which use both `skip` and `filter` options in the refactoring of this PR.

not-an-aardvark replied:
Personally, I think it would be better to just pick one behavior (e.g. don't count filtered tokens as skipped) and document the behavior, but not throw an error. We might as well allow an easy way to get the second comma.

mysticatea replied:
I added the note to documents.

mysticatea commented:
I just rebased this PR to resolve conflicts.
I think this is ready to review again.

ilyavolodin commented:
I'm still concern about 3% performance bump. Could you explain what this PR improves from the user perspective (other then slightly easier syntax for custom rule creation when it comes to tokens)? I'm wondering if the improvements are worth perf hit.

mysticatea commented:
As https://github.com/eslint/eslint/pull/7936/commits/e306942da9632a941d80260ba63a770eeb968958, we can write rules simpler. This would help contributors, plugin ecosystem, and power users which make custom rules.

Though `npm run perf` does not measure memory usage currently, this PR reduce 50% memory cost about TokenStore.

ilyavolodin commented:
Good point on memory usage vs. performance. I'll wait for others to jump in if they think that performance drop is too much. But if not - I'm good with this PR.

mysticatea commented on lib/rules/generator-star-spacing.js:
```diff
@@ -55,21 +55,26 @@ module.exports = {
+         * @param {ASTNode} node - The function node to get.
+         * @returns {Token} Found star token.
+         */
+        function getStartToken(node) {
```
Oops!

mysticatea commented on lib/token-store.js:
```diff
@@ -5,199 +5,1066 @@
+}
+
+/**
+ * Creates the cursor iterates tokens with options.
```
I tried to express an overload function.
I will add a note.

mysticatea commented on tests/lib/token-store.js:
```diff
@@ -89,6 +90,38 @@ describe("TokenStore", () => {
+
+        it("should retrieve matched tokens and comments in the node for root node with includeComments and filter option", () => {
+            check(
+                store.getTokens(Program, { includeComments: true, filter: t => t.type === "Block" }),
```
For the time being, I modified it with `filter: t => t.type.startsWith("Block")`

kaicataldo commented on lib/token-store.js:
```diff
@@ -5,199 +5,1071 @@
+    let filter = null;
+
+    if (typeof opts === "number") {
+        skip = opts | 0;
```
I've actually never personally used the bitwise OR operator. Just for my own understanding, is this being used for performance reasons (as opposed to `Math.floor()`, for instance)?

mysticatea replied:
Yes.
V8 can optimize integer (not double) in high level. Bitwise operator would tell V8 that this is integer.

.... Probably :smiley:

kaicataldo commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -261,23 +261,37 @@ module.exports = {
+* `getNodeByRangeIndex(index)` - returns the deepest node in the AST containing the given source index.
+
+> `skipOptions` is an object which has 3 properties; `skip`, `includeComments`, and `filter`. Default is `{skip: 0, includeComments: false, filter: null}`.
+> - The `skip` is a positive integer, the number of skipping tokens. If `filter` option is given at the same time, it doesn't count filtered tokens as skipped.
```
I had one one more thought on my second pass through this - do you think it's worth throwing an error with a helpful message when `skip` or `count` are not positive integers? I'm a fan of failing fast and showing a useful error message, since it can help people avoid frustrating debugging.

gyandeeps commented:
I will slowly review this PR. My first suggestion would be to create a folder `tokenStore` and then spplit the single 1000 line file into multiple files. Then have index.js file which exposes the main stuff from the folder. 
This is for readability. Thoughts @mysticatea ?

mysticatea commented:
- I split `token-store.js` to `token-store/*.js` files. Entry point is `token-store/index.js`.
- I added assertions for `options.skip` and `options.count`.
- I fixed plural in test descriptions.

platinumazure commented on lib/token-store/index.js:
```diff
@@ -0,0 +1,594 @@
+function createIndexMap(tokens, comments) {
+    const map = Object.create(null);
+    let it = 0;
+    let ic = 0;
```
Could these variables be renamed (e.g., tokenIndex, commentIndex)? (I think this came up earlier)

ilyavolodin commented:
I'm going to go ahead an merge this in so we can test it out before the next release.
